pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java

```diff
@@ -425,13 +426,14 @@ public SuccessResponse reloadSegment(
   /**
    * Helper method to find the existing table based on the given table name (with or without type suffix) and segment
    * name.
-   * TODO: Real-time table might also contain uploaded segments (not with LLC name), which is not handled here.
    */
   private String getExistingTable(String tableName, String segmentName) {
     TableType tableType = TableNameBuilder.getTableTypeFromTableName(tableName);
     if (tableType == null) {
       // Derive table type from segment name if the given table name doesn't have type suffix
-      tableType = LLCSegmentName.isLLCSegment(segmentName) ? TableType.REALTIME : TableType.OFFLINE;
```

Comment:
Is it safe to assume that `UploadedRealtimeSegmentName.isUploadedRealtimeSegmentName(segmentName)` will not return true for an offline table? I saw that it simply checks for the count of double underscores and returns true if there are 4 of them.

```
/**
   * Checks if the segment name is of the format: {prefix}__{tableName}__{partitionId}__{creationTime}__{suffix}
   * @param segmentName
   * @return boolean true if the segment name is of the format: {prefix}__{tableName}__{partitionId}__{creationTime}
   * __{suffix}
   */
  public static boolean isUploadedRealtimeSegmentName(String segmentName) {
    int numSeparators = 0;
    int index = 0;
    while ((index = segmentName.indexOf(SEPARATOR, index)) != -1) {
      numSeparators++;
      index += 2; // SEPARATOR.length()
    }
    return numSeparators == 4;
  }
```

Reply:
@ankitsultana, the implementation is similar to LLCSegment.isLLCSegment and we have table name restriction to not use `__`.
```
Preconditions.checkArgument(!tableName.contains(TABLE_NAME_FORBIDDEN_SUBSTRING),
        "'tableName' cannot contain double underscore ('__')");
```

It should be safe to assume this segment name will always mean a realtime table(based on javadocs). 

Additionally, I think we can add a restriction to only allow `UploadedRealtimeSegmentName` for realtime tables to make these kind of segment -> table type derivation safer.

Reply:
For LCCSegment too, it seems it is not guaranteed only for realtime table. There are no validations in `PinotSegmentUploadDownloadRestletResource` that LCCSegment should only be uploaded to realtime table and segment assignments for offline tables don't check segmentNames.

Additionally, there are places in code discouraging deriving table type from segmentName, I think we can either consolidate segmentName to table types in one place and/or enforce segmentNames to realtime table type:

```
// Fetch table name. Try to derive the table name from the parameter and then from segment metadata
      String rawTableName;
      if (StringUtils.isNotEmpty(tableName)) {
        rawTableName = TableNameBuilder.extractRawTableName(tableName);
      } else {
        // TODO: remove this when we completely deprecate the table name from segment metadata
        rawTableName = segmentMetadata.getTableName();
        LOGGER.warn("Table name is not provided as request query parameter when uploading segment: {} for table: {}",
            segmentName, rawTableName);
      }
```

Reply:
yeah, isLLCSegment and isUploadedRealtimeSegmentName are doing similar checks. And using segment name to derive table type is just best effort and not very reliable as far as I can tell. It's better to call the reloadSegment API with `tableNameWithType` as the type can be extracted this table name more reliably.

Reply:
@klsince do you recommend writing another API and deprecating this one (we cannot update this one for backward compatibility imo)? Or is it okay to use `isUploadedRealtimeSegmentName`?

Reply:
No, didn’t mean that. I see the current reloadSegment restful is already able to take tableNameWithType to extract table type. When table type can’t be retrieved that way, it falls back to derive type from segment name, which is not reliable even for LLC segment name anyway.

Reply:
Yeah but then should we start enforcing tableNameWithType or add this patch to fallback to extracting using segment name on a best effort basis? Worst case if segment naming is incorrect, we end up throwing TableNotFoundException which is okay i guess.

Long term maybe we should migrate all the APIs to a standard naming convention (using tableNameWithType always) but that seems to be a bigger alignment.